File: api/_lib/matcher.py
"""
매칭 엔진 모듈 (Vercel 서버리스용)

정산 시트와 계약 종료 시트 간 데이터를 매칭합니다.
1차 정확 매칭 → 퍼지 매칭 → 교차 검증
"""

import re
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Optional

import pandas as pd
from rapidfuzz import fuzz

# ============================================================
# 설정 상수
# ============================================================
FUZZY_THRESHOLD = 80
NORMALIZE_REMOVE_CHARS = [" ", "\t", "\n", "-", ".", "(", ")", "\u3000"]
DATE_FORMATS = [
    "%Y-%m-%d", "%Y.%m.%d", "%Y/%m/%d", "%Y%m%d",
    "%y-%m-%d", "%y.%m.%d", "%y/%m/%d",
]
# 종료 시트에서 계약자명이 비어있을 때 이름 폴백 컬럼
NAME_FALLBACK_COLUMNS_TERMINATION = ["주민 번호", "상호"]
NAME_FALLBACK_COLUMNS_SETTLEMENT = []
# 이미 종료 처리된 키워드
ALREADY_TERMINATED_KEYWORDS = ["계약종료", "해지", "종료"]
# 비고 메시지 템플릿
TERMINATION_NOTE_TEMPLATE = "계약종료 (만기일: {expiry_date})"


class MatchStatus(Enum):
    EXACT = "정확매칭"
    FUZZY = "유사매칭(수동확인)"
    UNMATCHED = "매칭실패"


@dataclass
class MatchResult:
    termination_index: int
    settlement_index: Optional[int]
    status: MatchStatus
    confidence: float
    match_details: str
    termination_row: dict = field(default_factory=dict)
    settlement_row: dict = field(default_factory=dict)
    verification_passed: bool = False


# ============================================================
# 유틸리티 함수
# ============================================================

def normalize_text(text: str) -> str:
    if pd.isna(text) or text is None:
        return ""
    text = str(text).strip()
    for char in NORMALIZE_REMOVE_CHARS:
        text = text.replace(char, "")
    return text.lower()


def normalize_name(name: str) -> str:
    """한글명/영문명 정규화"""
    if pd.isna(name) or name is None:
        return ""
    name = str(name).strip()
    if not name:
        return ""
    name = re.sub(r"\s+", " ", name)
    if re.match(r"^[가-힣\s]+$", name):
        return name.replace(" ", "")
    if re.match(r"^[a-zA-Z\s.\-]+$", name):
        name = name.replace(".", " ").replace("-", " ")
        return re.sub(r"\s+", " ", name).strip().lower()
    return name.replace(" ", "").lower()


def is_likely_name(text: str) -> bool:
    """주민번호 칸에 이름이 들어있는지 판별"""
    if pd.isna(text) or text is None:
        return False
    text = str(text).strip()
    if not text:
        return False
    if re.match(r"^[\d\-\s]+$", text):
        return False
    korean = re.sub(r"\s+", "", text)
    if re.match(r"^[가-힣]{2,5}$", korean):
        return True
    if re.match(r"^[a-zA-Z\s.\-]{2,}$", text) and any(c.isalpha() for c in text):
        return True
    return False


def normalize_business_number(biz_num: str) -> str:
    if pd.isna(biz_num) or biz_num is None:
        return ""
    return re.sub(r"[^0-9]", "", str(biz_num))


def parse_date(date_str: str) -> Optional[datetime]:
    if pd.isna(date_str) or date_str is None or str(date_str).strip() == "":
        return None
    date_str = str(date_str).strip()
    for fmt in DATE_FORMATS:
        try:
            return datetime.strptime(date_str, fmt)
        except ValueError:
            continue
    return None


def get_col(row: dict, col_name: str) -> str:
    """행에서 컬럼값을 유연하게 가져옴 (공백/줄바꿈 차이 무시)"""
    if col_name in row:
        v = row[col_name]
        return str(v) if not pd.isna(v) else ""
    norm = col_name.replace(" ", "").replace("\n", "").strip()
    for key in row.keys():
        if str(key).replace(" ", "").replace("\n", "").strip() == norm:
            v = row[key]
            return str(v) if not pd.isna(v) else ""
    return ""


def extract_name(row: dict, name_col: str, fallback_cols: list[str]) -> str:
    """계약자명 추출 (폴백: 주민번호칸 이름)"""
    name = get_col(row, name_col)
    if name.strip():
        return name.strip()
    for fb_col in fallback_cols:
        val = get_col(row, fb_col)
        if val.strip() and is_likely_name(val):
            return val.strip()
    return ""


# ============================================================
# 매칭 엔진
# ============================================================

def run_matching(
    settlement_df: pd.DataFrame,
    termination_df: pd.DataFrame,
    threshold: int = FUZZY_THRESHOLD,
) -> list[MatchResult]:
    """전체 매칭 실행, 결과 리스트 반환"""
    results = []
    matched_indices: set[int] = set()

    for t_idx, t_row in termination_df.iterrows():
        t_dict = t_row.to_dict()

        result = _exact_match(t_idx, t_dict, settlement_df, matched_indices)
        if result is None:
            result = _fuzzy_match(t_idx, t_dict, settlement_df, matched_indices, threshold)
        if result is None:
            result = MatchResult(
                termination_index=t_idx,
                settlement_index=None,
                status=MatchStatus.UNMATCHED,
                confidence=0.0,
                match_details="매칭되는 정산 건을 찾을 수 없음",
                termination_row=t_dict,
            )

        if result.settlement_index is not None:
            result = _cross_verify(result)
            matched_indices.add(result.settlement_index)

        results.append(result)

    return results


def _exact_match(t_idx, t_row, s_df, matched) -> Optional[MatchResult]:
    t_branch = normalize_text(get_col(t_row, "지점명"))
    t_room = normalize_text(get_col(t_row, "호실"))
    t_name_raw = extract_name(t_row, "계약자명", NAME_FALLBACK_COLUMNS_TERMINATION)
    t_name = normalize_name(t_name_raw)

    if not t_branch and not t_name:
        return None

    original = get_col(t_row, "계약자명")
    used_fb = (not original.strip()) and t_name_raw.strip()

    for s_idx, s_row in s_df.iterrows():
        if s_idx in matched:
            continue
        s_dict = s_row.to_dict()
        s_branch = normalize_text(get_col(s_dict, "지점명"))
        s_name_raw = extract_name(s_dict, "계약자명", NAME_FALLBACK_COLUMNS_SETTLEMENT)
        s_name = normalize_name(s_name_raw)
        s_room = normalize_text(get_col(s_dict, "호실"))

        fb_note = " (주민번호칸 이름)" if used_fb else ""

        if t_branch == s_branch and t_name == s_name and t_room == s_room:
            return MatchResult(
                termination_index=t_idx, settlement_index=s_idx,
                status=MatchStatus.EXACT, confidence=100.0,
                match_details=f"정확매칭: {t_branch}/{t_name_raw}{fb_note}/{t_room}",
                termination_row=t_row, settlement_row=s_dict,
            )
        if t_branch == s_branch and t_name == s_name and (not t_room or not s_room):
            return MatchResult(
                termination_index=t_idx, settlement_index=s_idx,
                status=MatchStatus.EXACT, confidence=90.0,
                match_details=f"정확매칭(호실미확인): {t_branch}/{t_name_raw}{fb_note}",
                termination_row=t_row, settlement_row=s_dict,
            )
    return None


def _fuzzy_match(t_idx, t_row, s_df, matched, threshold) -> Optional[MatchResult]:
    t_branch = get_col(t_row, "지점명")
    t_name_raw = extract_name(t_row, "계약자명", NAME_FALLBACK_COLUMNS_TERMINATION)
    t_room = get_col(t_row, "호실")

    if not t_branch.strip() and not t_name_raw.strip():
        return None

    original = get_col(t_row, "계약자명")
    used_fb = (not original.strip()) and t_name_raw.strip()

    best_score, best_idx, best_row, best_detail = 0.0, None, None, ""

    for s_idx, s_row in s_df.iterrows():
        if s_idx in matched:
            continue
        s_dict = s_row.to_dict()
        s_branch = get_col(s_dict, "지점명")
        s_name_raw = extract_name(s_dict, "계약자명", NAME_FALLBACK_COLUMNS_SETTLEMENT)
        s_room = get_col(s_dict, "호실")

        br = fuzz.ratio(normalize_text(t_branch), normalize_text(s_branch)) if t_branch.strip() and s_branch.strip() else 0
        tn, sn = normalize_name(t_name_raw), normalize_name(s_name_raw)
        cn = max(fuzz.ratio(tn, sn), fuzz.token_sort_ratio(tn, sn)) if tn and sn else 0
        rm = fuzz.ratio(normalize_text(t_room), normalize_text(s_room)) if t_room.strip() and s_room.strip() else 100

        total = br * 0.3 + cn * 0.5 + rm * 0.2
        if total > best_score:
            best_score, best_idx, best_row = total, s_idx, s_dict
            fb = " [이름폴백]" if used_fb else ""
            best_detail = f"유사매칭: 지점={br:.0f}%, 계약자={cn:.0f}%{fb}, 호실={rm:.0f}% → {total:.1f}%"

    if best_score >= threshold and best_idx is not None:
        return MatchResult(
            termination_index=t_idx, settlement_index=best_idx,
            status=MatchStatus.FUZZY, confidence=best_score,
            match_details=best_detail,
            termination_row=t_row, settlement_row=best_row,
        )
    return None


def _cross_verify(result: MatchResult) -> MatchResult:
    verified = []
    t_biz = normalize_business_number(get_col(result.termination_row, "사업자 등록번호"))
    s_biz = normalize_business_number(get_col(result.settlement_row, "사업자등록번호"))
    if t_biz and s_biz:
        verified.append("사업자번호일치" if t_biz == s_biz else "사업자번호불일치")

    t_date = parse_date(get_col(result.termination_row, "계약 시작 날짜"))
    s_date = parse_date(get_col(result.settlement_row, "계약시작일자"))
    if t_date and s_date:
        verified.append("계약시작일일치" if t_date == s_date else "계약시작일불일치")

    has_ok = any("일치" in v and "불일치" not in v for v in verified)
    has_ng = any("불일치" in v for v in verified)

    if has_ok and not has_ng:
        result.verification_passed = True
        result.match_details += f" [검증통과: {','.join(verified)}]"
    elif has_ng:
        result.verification_passed = False
        result.match_details += f" [검증실패: {','.join(verified)}]"
        if result.status == MatchStatus.EXACT:
            result.status = MatchStatus.FUZZY
    else:
        result.verification_passed = True
        result.match_details += " [검증데이터없음]"

    return result


# ============================================================
# 결과 → JSON 직렬화
# ============================================================

def results_to_json(results: list[MatchResult]) -> list[dict]:
    """MatchResult 리스트를 JSON-serializable dict 리스트로 변환"""
    out = []
    for r in results:
        out.append({
            "termination_index": r.termination_index,
            "settlement_index": r.settlement_index,
            "status": r.status.value,
            "confidence": round(r.confidence, 1),
            "match_details": r.match_details,
            "verification_passed": r.verification_passed,
            "t_branch": get_col(r.termination_row, "지점명"),
            "t_contractor": extract_name(r.termination_row, "계약자명", NAME_FALLBACK_COLUMNS_TERMINATION),
            "t_room": get_col(r.termination_row, "호실"),
            "t_expiry": get_col(r.termination_row, "계약 만기 날짜"),
            "t_biz_num": get_col(r.termination_row, "사업자 등록번호"),
            "s_branch": get_col(r.settlement_row, "지점명") if r.settlement_row else "",
            "s_contractor": get_col(r.settlement_row, "계약자명") if r.settlement_row else "",
            "s_room": get_col(r.settlement_row, "호실") if r.settlement_row else "",
            "s_biz_num": get_col(r.settlement_row, "사업자등록번호") if r.settlement_row else "",
            "s_note": get_col(r.settlement_row, "계약변경/해지/비고") if r.settlement_row else "",
        })
    return out


def summary(results: list[MatchResult]) -> dict:
    """매칭 결과 요약"""
    total = len(results)
    exact = sum(1 for r in results if r.status == MatchStatus.EXACT)
    fuzzy_ = sum(1 for r in results if r.status == MatchStatus.FUZZY)
    unmatched = sum(1 for r in results if r.status == MatchStatus.UNMATCHED)
    return {
        "total": total,
        "exact": exact,
        "fuzzy": fuzzy_,
        "unmatched": unmatched,
    }

File: web/lib/matcher.js
/**
 * 매칭 엔진 모듈
 *
 * 정산 시트와 계약 종료 시트 간 데이터를 매칭합니다.
 * 1차 정확 매칭 → 2차 퍼지 매칭 → 교차 검증
 */

const fuzz = require("fuzzball");

// ============================================================
// 설정
// ============================================================

/** 매칭 키 (종료 시트 컬럼 → 정산 시트 컬럼) */
const MATCHING_KEYS = {
  지점명: { termination: "지점명", settlement: "지점명" },
  계약자명: { termination: "계약자명", settlement: "계약자명" },
  호실: { termination: "호실", settlement: "호실" },
};

/** 교차 검증 키 */
const VERIFY_KEYS = {
  사업자등록번호: { termination: "사업자 등록번호", settlement: "사업자등록번호" },
  계약시작일자: { termination: "계약 시작 날짜", settlement: "계약시작일자" },
};

/** 동기화 컬럼 */
const SYNC_KEYS = {
  계약만기날짜: { termination: "계약 만기 날짜", settlement: "계약종료일자" },
  결제금액: { termination: "결제 금액", settlement: "결제금액" },
};

/** 이름 폴백 컬럼 (계약자명이 비어있을 때 확인할 컬럼) */
const NAME_FALLBACK_COLUMNS = ["주민 번호", "주민번호", "상호"];

/** 이미 종료 처리된 키워드 */
const TERMINATED_KEYWORDS = ["계약종료", "해지", "종료"];

/** 정규화 시 제거할 문자 */
const REMOVE_CHARS = [" ", "\t", "\n", "-", ".", "(", ")", "\u3000"];

// ============================================================
// 유틸 함수
// ============================================================

/**
 * 행에서 컬럼 값을 유연하게 가져옵니다.
 * 컬럼명의 공백/줄바꿈 차이를 허용합니다.
 */
function getVal(row, colName) {
  if (!row || !colName) return "";
  // 정확한 키
  if (row[colName] !== undefined && row[colName] !== null) {
    return String(row[colName]).trim();
  }
  // 공백/줄바꿈 제거 후 매칭
  const norm = colName.replace(/\s+/g, "");
  for (const key of Object.keys(row)) {
    if (key.replace(/\s+/g, "") === norm) {
      return row[key] !== undefined && row[key] !== null
        ? String(row[key]).trim()
        : "";
    }
  }
  return "";
}

/** 텍스트 정규화 (공백, 특수문자 제거, 소문자) */
function normalizeText(text) {
  if (!text) return "";
  let t = String(text).trim();
  for (const ch of REMOVE_CHARS) {
    t = t.split(ch).join("");
  }
  return t.toLowerCase();
}

/** 주민번호가 아니라 이름인지 판별 */
function isLikelyName(text) {
  if (!text || !String(text).trim()) return false;
  const t = String(text).trim();
  // 숫자+하이픈만 → 주민번호
  if (/^[\d\-\s]+$/.test(t)) return false;
  // 한글 2~5자
  if (/^[가-힣]{2,5}$/.test(t.replace(/\s/g, ""))) return true;
  // 영문 이름
  if (/^[a-zA-Z\s.\-]{2,}$/.test(t) && /[a-zA-Z]/.test(t)) return true;
  // 한글+영문 혼합
  if (/[가-힣]/.test(t) && /[a-zA-Z]/.test(t)) return true;
  return false;
}

/** 이름 정규화 (한글: 공백제거, 영문: 소문자+공백통일) */
function normalizeName(name) {
  if (!name) return "";
  let n = String(name).trim().replace(/\s+/g, " ");
  // 한글만
  if (/^[가-힣\s]+$/.test(n)) return n.replace(/\s/g, "");
  // 영문만
  if (/^[a-zA-Z\s.\-]+$/.test(n)) {
    return n.replace(/[.\-]/g, " ").replace(/\s+/g, " ").trim().toLowerCase();
  }
  // 혼합
  return n.replace(/\s/g, "").toLowerCase();
}

/** 계약자명 추출 (폴백: 주민번호 칸에서 이름 탐색) */
function extractContractorName(row, colName) {
  const name = getVal(row, colName);
  if (name) return name;
  for (const fb of NAME_FALLBACK_COLUMNS) {
    const val = getVal(row, fb);
    if (val && isLikelyName(val)) return val;
  }
  return "";
}

/** 사업자등록번호 정규화 (숫자만) */
function normalizeBizNum(num) {
  if (!num) return "";
  return String(num).replace(/[^0-9]/g, "");
}

/** 날짜 문자열 정규화 (YYYY-MM-DD 형태로 통일) */
function normalizeDate(dateStr) {
  if (!dateStr || !String(dateStr).trim()) return "";
  const d = String(dateStr).trim();
  // 다양한 형식 시도
  const formats = [
    /^(\d{4})[-./](\d{1,2})[-./](\d{1,2})$/,
    /^(\d{2})[-./](\d{1,2})[-./](\d{1,2})$/,
  ];
  for (const fmt of formats) {
    const m = d.match(fmt);
    if (m) {
      let year = m[1].length === 2 ? "20" + m[1] : m[1];
      return `${year}-${m[2].padStart(2, "0")}-${m[3].padStart(2, "0")}`;
    }
  }
  return d;
}

// ============================================================
// 매칭 엔진
// ============================================================

/**
 * 메인 매칭 함수
 * @param {Object[]} settlementRows - 정산 시트 행 배열
 * @param {Object[]} terminationRows - 종료 시트 행 배열
 * @param {number} threshold - 퍼지 매칭 임계값 (기본 80)
 * @returns {Object[]} 매칭 결과 배열
 */
function matchContracts(settlementRows, terminationRows, threshold = 80) {
  const results = [];
  const matchedSettlementIndices = new Set();

  for (let tIdx = 0; tIdx < terminationRows.length; tIdx++) {
    const tRow = terminationRows[tIdx];

    // 1단계: 정확 매칭
    let result = exactMatch(tRow, tIdx, settlementRows, matchedSettlementIndices);

    // 2단계: 퍼지 매칭
    if (!result) {
      result = fuzzyMatch(tRow, tIdx, settlementRows, matchedSettlementIndices, threshold);
    }

    // 매칭 실패
    if (!result) {
      result = {
        terminationIndex: tIdx,
        settlementIndex: null,
        status: "매칭실패",
        confidence: 0,
        details: "매칭되는 정산 건을 찾을 수 없음",
        terminationRow: tRow,
        settlementRow: null,
        verified: false,
      };
    }

    // 3단계: 교차 검증
    if (result.settlementIndex !== null) {
      result = crossVerify(result);
      matchedSettlementIndices.add(result.settlementIndex);
    }

    results.push(result);
  }

  return results;
}

/** 정확 매칭 */
function exactMatch(tRow, tIdx, sRows, matched) {
  const tBranch = normalizeText(getVal(tRow, MATCHING_KEYS.지점명.termination));
  const tName = normalizeName(extractContractorName(tRow, MATCHING_KEYS.계약자명.termination));
  const tRoom = normalizeText(getVal(tRow, MATCHING_KEYS.호실.termination));

  if (!tBranch && !tName) return null;

  const originalName = getVal(tRow, MATCHING_KEYS.계약자명.termination);
  const usedFallback = !originalName && tName;

  for (let sIdx = 0; sIdx < sRows.length; sIdx++) {
    if (matched.has(sIdx)) continue;
    const sRow = sRows[sIdx];

    const sBranch = normalizeText(getVal(sRow, MATCHING_KEYS.지점명.settlement));
    const sName = normalizeName(extractContractorName(sRow, MATCHING_KEYS.계약자명.settlement));
    const sRoom = normalizeText(getVal(sRow, MATCHING_KEYS.호실.settlement));

    const fbNote = usedFallback ? " (주민번호칸에서 이름추출)" : "";

    // 3가지 모두 일치
    if (tBranch === sBranch && tName === sName && tRoom === sRoom) {
      return {
        terminationIndex: tIdx,
        settlementIndex: sIdx,
        status: "정확매칭",
        confidence: 100,
        details: `지점=${tBranch}, 계약자=${tName}${fbNote}, 호실=${tRoom}`,
        terminationRow: tRow,
        settlementRow: sRow,
        verified: false,
      };
    }

    // 지점 + 이름 일치 (호실 비어있음)
    if (tBranch === sBranch && tName === sName && (!tRoom || !sRoom)) {
      return {
        terminationIndex: tIdx,
        settlementIndex: sIdx,
        status: "정확매칭",
        confidence: 90,
        details: `지점=${tBranch}, 계약자=${tName}${fbNote} (호실미확인)`,
        terminationRow: tRow,
        settlementRow: sRow,
        verified: false,
      };
    }
  }
  return null;
}

/** 퍼지 매칭 */
function fuzzyMatch(tRow, tIdx, sRows, matched, threshold) {
  const tBranch = getVal(tRow, MATCHING_KEYS.지점명.termination);
  const tNameRaw = extractContractorName(tRow, MATCHING_KEYS.계약자명.termination);
  const tRoom = getVal(tRow, MATCHING_KEYS.호실.termination);

  if (!tBranch.trim() && !tNameRaw.trim()) return null;

  const originalName = getVal(tRow, MATCHING_KEYS.계약자명.termination);
  const usedFallback = !originalName.trim() && tNameRaw.trim();

  let bestScore = 0;
  let bestIdx = null;
  let bestRow = null;
  let bestDetail = "";

  for (let sIdx = 0; sIdx < sRows.length; sIdx++) {
    if (matched.has(sIdx)) continue;
    const sRow = sRows[sIdx];

    const sBranch = getVal(sRow, MATCHING_KEYS.지점명.settlement);
    const sNameRaw = extractContractorName(sRow, MATCHING_KEYS.계약자명.settlement);
    const sRoom = getVal(sRow, MATCHING_KEYS.호실.settlement);

    // 지점명 유사도
    const branchScore =
      tBranch.trim() && sBranch.trim()
        ? fuzz.ratio(normalizeText(tBranch), normalizeText(sBranch))
        : 0;

    // 계약자명 유사도 (ratio + token_sort_ratio 중 높은 값)
    const tNameN = normalizeName(tNameRaw);
    const sNameN = normalizeName(sNameRaw);
    let nameScore = 0;
    if (tNameN && sNameN) {
      nameScore = Math.max(
        fuzz.ratio(tNameN, sNameN),
        fuzz.token_sort_ratio(tNameN, sNameN)
      );
    }

    // 호실 유사도
    const roomScore =
      tRoom.trim() && sRoom.trim()
        ? fuzz.ratio(normalizeText(tRoom), normalizeText(sRoom))
        : 100;

    // 가중 평균 (지점30% + 이름50% + 호실20%)
    const total = branchScore * 0.3 + nameScore * 0.5 + roomScore * 0.2;

    if (total > bestScore) {
      bestScore = total;
      bestIdx = sIdx;
      bestRow = sRow;
      const fbNote = usedFallback ? " [이름폴백]" : "";
      bestDetail = `지점=${branchScore.toFixed(0)}%, 계약자=${nameScore.toFixed(0)}%${fbNote}, 호실=${roomScore.toFixed(0)}% → ${total.toFixed(1)}%`;
    }
  }

  if (bestScore >= threshold && bestIdx !== null) {
    return {
      terminationIndex: tIdx,
      settlementIndex: bestIdx,
      status: "유사매칭",
      confidence: Math.round(bestScore * 10) / 10,
      details: bestDetail,
      terminationRow: tRow,
      settlementRow: bestRow,
      verified: false,
    };
  }
  return null;
}

/** 교차 검증 */
function crossVerify(result) {
  const checks = [];

  // 사업자등록번호
  const tBiz = normalizeBizNum(getVal(result.terminationRow, VERIFY_KEYS.사업자등록번호.termination));
  const sBiz = normalizeBizNum(getVal(result.settlementRow, VERIFY_KEYS.사업자등록번호.settlement));
  if (tBiz && sBiz) {
    checks.push(tBiz === sBiz ? "사업자번호일치" : "사업자번호불일치");
  }

  // 계약시작일자
  const tDate = normalizeDate(getVal(result.terminationRow, VERIFY_KEYS.계약시작일자.termination));
  const sDate = normalizeDate(getVal(result.settlementRow, VERIFY_KEYS.계약시작일자.settlement));
  if (tDate && sDate) {
    checks.push(tDate === sDate ? "계약시작일일치" : "계약시작일불일치");
  }

  const hasMatch = checks.some((c) => c.includes("일치") && !c.includes("불일치"));
  const hasMismatch = checks.some((c) => c.includes("불일치"));

  if (hasMatch && !hasMismatch) {
    result.verified = true;
    result.details += ` [검증통과: ${checks.join(", ")}]`;
  } else if (hasMismatch) {
    result.verified = false;
    result.details += ` [검증실패: ${checks.join(", ")}]`;
    if (result.status === "정확매칭") {
      result.status = "유사매칭";
      result.details += " → 수동확인 필요";
    }
  } else {
    result.verified = true;
    result.details += " [검증데이터없음]";
  }

  return result;
}

/** 이미 종료 처리된 건인지 확인 */
function isAlreadyTerminated(noteValue) {
  if (!noteValue) return false;
  return TERMINATED_KEYWORDS.some((kw) => noteValue.includes(kw));
}

module.exports = {
  matchContracts,
  getVal,
  isAlreadyTerminated,
  MATCHING_KEYS,
  VERIFY_KEYS,
  SYNC_KEYS,
  TERMINATED_KEYWORDS,
};

